vignettes/data.Rmd

```diff
@@ -0,0 +1,186 @@
+
+* describing the data (here it will be a single numeric value, but the data could be arbitrarily complicated)
+* implementing a method to do the comparison
+* describe how to marshal the data into this structure
+
+```{r, echo = FALSE, results = "asis"}
+cc_output(readLines(system.file("examples/volatility.cpp", package = "dust")))
+```
+
+The first part, the data definition, is the definition
+
+```cc
+  struct data_t {
+    real_t observed;
+  };
+```
+
+which replaces the usual
+
+```cc
+  typedef dust::no_data data_t;
+```
+
+This structure can contain whatever you want, including things like a `std::vector` of values. In our use we've typically only had `real_t` and `int` though, even for complex models.
+
+The comparison is implemented by the method `compare_data`
+
+```cc
+  real_t compare_data(const real_t * state, const data_t& data,
+                      dust::rng_state_t<real_t>& rng_state) {
+    return dust::dnorm(data.observed, shared->gamma * state[0],
+                       shared->tau, true);
+  }
+```
+
+This looks very much like the R version above:
+
+* `dnorm` is replaced with the dust function `dust::dnorm` (do not use R API functions here, as this will be evaluated in a multi-threaded context)
+* the random number generator is available here, if your comparison is itself stochastic
+* unlike the R version where the comparison returns a vector across all particles, the C++ comparison is done for each particle
+
+Finally, the data marshalling is done by the `dust_data` template, within the `dust` namespace
+
+```cc
+namespace dust {
+template <>
+volatility::data_t dust_data<volatility>(cpp11::list data) {
+  return volatility::data_t{cpp11::as_cpp<double>(data["observed"])};
+}
+}
+```
+
+Here, you can use any function you could use from `cpp11`, much like within the `dust_pars` template specialisation. The input will be a list, corresponding to the data *at a single time point*.  The data.frame above will first be processed with `dust::dust_data`, so the first few entries look like:
+
+```{r}
+head(dust::dust_data(data), 3)
+```
+
+This is definitely a bit fiddly!  If using `odin.dust`, then this would be somewhat simplified as you could provide a single C++ file containing something like
+
+```cc
+// [[odin.dust::compare_data(observed = real_t)]]
+// [[odin.dust::compare_function]]
+template <typename T>
+typename T::real_t compare(const typename T::real_t * state,
+                           const typename T::data_t& data,
+                           const typename T::internal_t internal,
+                           std::shared_ptr<const typename T::shared_t> shared,
+                           dust::rng_state_t<typename T::real_t>& rng_state) {
+  return dust::dnorm(data.observed, odin(gamma) * odin(value), odin(tau), true);
+}
+```
+
+and the correct code would be generated.
+
+Once set up, the compare function can be used. First, create the model
```

Comment:
Maybe also point to some examples in the code for other cases for users to look at (I think one of the SIR models has a compare function)